pkg/expr/classic/reduce.go

```diff
@@ -0,0 +1,205 @@
+package classic
+
+import (
+	"math"
+	"sort"
+
+	"github.com/grafana/grafana-plugin-sdk-go/data"
+	"github.com/grafana/grafana/pkg/expr/mathexp"
+)
+
+func nilOrNaN(f *float64) bool {
+	return f == nil || math.IsNaN(*f)
+}
+
+func (cr classicReducer) ValidReduceFunc() bool {
+	switch cr {
+	case "avg", "sum", "min", "max", "count", "last", "median":
+		return true
+	case "diff", "diff_abs", "percent_diff", "percent_diff_abs", "count_not_null":
+		return true
+	}
+	return false
+}
+
+//nolint: gocyclo
+func (cr classicReducer) Reduce(series mathexp.Series) mathexp.Number {
+	num := mathexp.NewNumber("", nil)
+	num.SetValue(nil)
+
+	if series.Len() == 0 {
+		return num
+	}
+
+	value := float64(0)
+	allNull := true
+
+	vF := series.Frame.Fields[series.ValueIdx]
+
+	switch cr {
+	case "avg":
+		validPointsCount := 0
+		for i := 0; i < vF.Len(); i++ {
+			if f, ok := vF.At(i).(*float64); ok {
+				if nilOrNaN(f) {
+					continue
+				}
+				value += *f
+				validPointsCount++
+				allNull = false
+			}
+		}
+		if validPointsCount > 0 {
+			value /= float64(validPointsCount)
+		}
+	case "sum":
+		for i := 0; i < vF.Len(); i++ {
+			if f, ok := vF.At(i).(*float64); ok {
+				if nilOrNaN(f) {
+					continue
+				}
+				value += *f
+				allNull = false
+			}
+		}
+	case "min":
+		value = math.MaxFloat64
+		for i := 0; i < vF.Len(); i++ {
+			if f, ok := vF.At(i).(*float64); ok {
+				if nilOrNaN(f) {
+					continue
+				}
+				allNull = false
+				if value > *f {
+					value = *f
+				}
+			}
+		}
+		if allNull {
```

Comment:
Is that required? I think that it will fall in that case anyway:
https://github.com/grafana/grafana/blob/cb6af9bf4f828816065f4bb2d1196ab8f7d91929/pkg/expr/classic/reduce.go#L153-L155

Reply:
This is the equiv of "no data", and value is initialized to 0, so need the pointer in the case.

A bit strange, but porting existing code and trying to keep it close.